Modules/posixmodule.c

```diff
@@ -5088,384 +5088,358 @@ os__path_splitroot_impl(PyObject *module, path_t *path)
-            close_file = FALSE;
-        }
-        else {
-            hfile = CreateFileW(_path.wide, FILE_READ_ATTRIBUTES, 0, NULL,
-                                OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+    else if (_path->wide) {
+        result = _testFileExistsByName(_path->wide, followLinks);
+    }
+    Py_END_ALLOW_THREADS
+
+    path_cleanup(_path);
+    return result;
+}
+
+
+static int
+_testFileType(path_t *_path, PyObject *path, int testedType)
+{
+    BOOL result = FALSE;
+    if (!path_converter(path, _path)) {
+        path_cleanup(_path);
+        if (PyErr_ExceptionMatches(PyExc_ValueError)) {
+            PyErr_Clear();
+            return FALSE;
         }
+        return -1;
+    }
+
+    Py_BEGIN_ALLOW_THREADS
+    if (_path->fd != -1) {
+        HANDLE hfile = _Py_get_osfhandle_noraise(_path->fd);
         if (hfile != INVALID_HANDLE_VALUE) {
-            if (GetFileInformationByHandleEx(hfile, FileBasicInfo, &info,
-                                            sizeof(info)))
-            {
-                result = !(info.FileAttributes & FILE_ATTRIBUTE_DIRECTORY);
-            }
-            else {
-                result = 0;
-            }
-            if (close_file) {
-                CloseHandle(hfile);
-            }
-        }
-        else {
-            STRUCT_STAT st;
-            switch (GetLastError()) {
-            case ERROR_ACCESS_DENIED:
-            case ERROR_SHARING_VIOLATION:
-            case ERROR_CANT_ACCESS_FILE:
-            case ERROR_INVALID_PARAMETER:
-                if (STAT(_path.wide, &st)) {
-                    result = 0;
-                }
-                else {
-                    result = S_ISREG(st.st_mode);
-                }
-                break;
-            default:
-                result = 0;
-            }
+            result = _testFileTypeByHandle(hfile, testedType, TRUE);
         }
     }
+    else if (_path->wide) {
+        result = _testFileTypeByName(_path->wide, testedType);
+    }
     Py_END_ALLOW_THREADS
 
-    path_cleanup(&_path);
-    if (result) {
-        Py_RETURN_TRUE;
-    }
-    Py_RETURN_FALSE;
+    path_cleanup(_path);
+    return result;
 }
 
 
 /*[clinic input]
-os._path_exists
+os._path_exists -> bool
 
-    path: 'O'
+    path: object
+    /
 
-Test whether a path exists.  Returns False for broken symbolic links
+Test whether a path exists.  Returns False for broken symbolic links.
 
 [clinic start generated code]*/
 
-static PyObject *
+static int
 os__path_exists_impl(PyObject *module, PyObject *path)
-/*[clinic end generated code: output=f508c3b35e13a249 input=380f77cdfa0f7ae8]*/
+/*[clinic end generated code: output=8f784b3abf9f8588 input=2777da15bc4ba5a3]*/
 {
-    HANDLE hfile;
-    BOOL close_file = TRUE;
-    path_t _path = PATH_T_INITIALIZE("exists", "path", 0, 1);
-    int result;
-    BOOL slow_path = TRUE;
-    FILE_STAT_BASIC_INFORMATION statInfo;
+    path_t _path = PATH_T_INITIALIZE("_path_exists", "path", 0, 1);
```

Comment:
I'm not super thrilled about using this directly, even though I know we were before. Any chance that adding an extra argument to the `path_t` definition to make it not raise `ValueError` would be easy? I haven't looked at it myself.

Reply:
That's precisely what we'll be doing after this pull request: #118355.